pkg/query-service/app/logs/v3/query_builder.go

```diff
@@ -371,8 +402,28 @@ func addOffsetToQuery(query string, offset uint64) string {
 	return fmt.Sprintf("%s OFFSET %d", query, offset)
 }
 
-func PrepareLogsQuery(start, end int64, queryType v3.QueryType, panelType v3.PanelType, mq *v3.BuilderQuery) (string, error) {
-	query, err := buildLogsQuery(panelType, start, end, mq.StepInterval, mq)
+func PrepareLogsQuery(start, end int64, queryType v3.QueryType, panelType v3.PanelType, mq *v3.BuilderQuery, graphLimitQtype string) (string, error) {
+
+	if graphLimitQtype == constants.FirstQueryGraphLimit {
+		// give me just the groupby names
+		query, err := buildLogsQuery(panelType, start, end, mq.StepInterval, mq, graphLimitQtype)
+		if err != nil {
+			return "", err
+		}
+		query = addLimitToQuery(query, mq.Limit)
+
+		return query, nil
+	} else if graphLimitQtype == constants.SecondQueryGraphLimit {
+		query, err := buildLogsQuery(panelType, start, end, mq.StepInterval, mq, graphLimitQtype)
+		if err != nil {
+			return "", err
+		}
+		return query, nil
+	}
```

Comment:
Okay, but would you suggest doing it in this PR ? I beleive this will add a good amount of changes.

Reply:
I agree on updating query builder for panel type. But that will be a complete rewrite of query builder and that can be taken up separately.

Reply:
I won't be able to provide any good feedback on this as is since I can't wrap my head around it. Speaking from experience, the _taken up separately_ doesn't get picked up separately as people move on to different things. I didn't mean to say it in this PR. I would first create refactor PR and then rebase this when that gets merged.

Reply:
I also don't mind approving this, but I believe that's pointless and can't give any meaningful reviews for future changes if I don't understand things.

Reply:
These changes have to be taken up separately **seriously** as in the future it's going to be difficult making changes and introducing new panels without separation of panels in codebase.

Yes, PR can't be approved if we don't understand things. Maybe we can get on a call with @nityanandagohain for this PR to understand this PR and clear doubts.